app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.appium.uiautomator2.server.mjpeg;
+
+import android.app.UiAutomation;
+import android.graphics.Bitmap;
+
+import java.lang.Math;
+import java.util.List;
+import java.util.Iterator;
+import java.util.Locale;
+
+import io.appium.uiautomator2.model.internal.CustomUiDevice;
+import io.appium.uiautomator2.server.ServerConfig;
+import io.appium.uiautomator2.utils.ScreenshotHelper;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public class MjpegScreenshotStream extends Thread {
+    private static final UiAutomation uia =
+        CustomUiDevice.getInstance().getInstrumentation().getUiAutomation();
+    private final List<MjpegScreenshotClient> clients;
+    private boolean stopped = false;
+
+    MjpegScreenshotStream(List<MjpegScreenshotClient> clients) {
+        this.clients = clients;
+    }
+
+    @Override
+    public void interrupt() {
+        this.stopped = true;
+        super.interrupt();
+    }
+
+    @Override
+    public void run() {
+        while (!stopped) {
+            if (clients.size() == 0) {
+                try {
+                    sleep(500);
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+                continue;
+            }
+
+            // how long each loop should take in milliseconds
+            long targetInterval =
```

Comment:
consider switching to ScheduledThreadPoolExecutor -> scheduleAtFixedRate

Reply:
With the changed made to synchronization, short-circuiting / minimizing busy-loops, I think we're in pretty good shape.